### PR TITLE
Document support for ssh over SOCKS5

### LINKED ISSUE #17754: Socks5 proxy support
Hi, I'm wondering whether you're planning to add support for socks5 proxy in terraform like Packer has?

It would be very useful to have this feature cause right now I'm struggling with running terraform behind our company's proxy and have to use it through ssh tunnel and bastion host.

I was trying to add socks5 support by myself but faced several issues:
1. I added new connection properties for proxy to connectionInfo struct in "communicator/ssh/provisioner.go"(I suppose it is the right place for modification)
```
type connectionInfo struct {
/*
  some other properties here
*/
SSHProxyHost      string `mapstructure:"proxy_host"`
SSHProxyPort      int    `mapstructure:"proxy_port"`
SSHProxyUser      string `mapstructure:"proxy_user"`
SSHProxyPassword  string `mapstructure:"proxy_password"`
}
```
 but when I built a binary I got this error -

```
Error: aws_instance.example: unknown 'connection' argument "proxy_host"
```

**So it looks like terraform can't get new properties from configuration.**

I also added almost identical code from packer source to "communicator/ssh/communicator.go"
```
// ProxyConnectFunc func is a new method for returning a function
// that connects to the host using SOCKS5 proxy like Packer does
func ProxyConnectFunc(socksProxy string, socksAuth *proxy.Auth, network, addr string) func() (net.Conn, error) {
	return func() (net.Conn, error) {
		dialer, err := proxy.SOCKS5("tcp", socksProxy, socksAuth, proxy.Direct)
		if err != nil {
			return nil, fmt.Errorf("Can't connect to the proxy: %s", err)
		}

		c, err := dialer.Dial(network, addr)
		if err != nil {
			return nil, err
		}

		return c, nil
	}
}
```

### PR DESCRIPTION
**Pull request by sorenisanerd**
Fixes #17754, surprisingly.

`golang.org/x/net/proxy.FromURL` (which Terraform uses to add support for HTTP and HTTPS) supports socks5 out of the box:

https://cs.opensource.google/go/x/net/+/refs/tags/v0.15.0:proxy/proxy.go;l=91

### CODE REVIEW

**Comment by crw** [comment]
Thanks @sorenisanerd for this update and spotting that this functionality now exists. I plan to bring this to triage and hopefully get it reviewed, would you please sign the CLA so that we can use this submission? Thanks again!

**Comment by crw** [comment]
Hi @sorenisanerd, we would be happy to approve these changes once you sign the CLA. Once you have done that, I'll let the docs team do a once-over on the style and we will be set. Thanks!

**Comment by crw** [comment]
Also, I wonder if it is worth updating to mention `socks5h` support? To be honest, I am not clear if that is fully supported by the new functionality or just recognized as a keyword that defaults to some more basic socks5 functionality (just because it is caught by the same case statement in that code snippet, above). I leave it to your judgement. Thanks!

File: website/docs/language/resources/provisioners/connection.mdx
Comment (trujillo-adam) [inline website/docs/language/resources/provisioners/connection.mdx:116]:
```suggestion
## Connection through HTTP and SOCKS5 proxies with SSH
```
Per our style guide, we should only use slashes as part of code snippets.

File: website/docs/language/resources/provisioners/connection.mdx
Comment (trujillo-adam) [inline website/docs/language/resources/provisioners/connection.mdx:122]:
```suggestion
| `proxy_scheme` | You can specify one of the following values: `http`, `https`, `socks5` | |
```
I _think_ the intention is to describe what values to set for this param. If not, then HTTP and HTTPS should be uppercase because they are proper acronyms. SOCKS5 also appears to be spelled in all caps, but that appears to be for branding purposes.

**Review by trujillo-adam — COMMENTED** [review]
Thanks for adding this - just a few edits for consistency with our styles

File: website/docs/language/resources/provisioners/connection.mdx
Comment (crw) [reply website/docs/language/resources/provisioners/connection.mdx:122]: Correct these are values.

**Comment by crw** [comment]
Hi @sorenisanerd, the next steps for this PR are to accept the CLA and the accept changes proposed above or discuss the changes. Thanks!

**Comment by sorenisanerd** [comment]
Yep, still working on the CLA.

**Comment by sorenisanerd** [comment]
@crw, I signed the CLA. Apologies for the delay.

**Comment by crw** [comment]
@sorenisanerd thank you! The next and potentially final step is to accept the suggested changes above, or discuss the changes if you disagree with them. Thanks!

**Comment by sorenisanerd** [comment]
Done-diddly-done.

**Review by crw — APPROVED** [review]

**Pull request merged**
